Fix interrupts handling

wipawel opened the pull request:
Up to now the interrupts for timers and keyboard were completely broken - especially when happen in usermode.
This series attempts to toss out most obvious bugs and does some cleanup.

wipawel committed:
setup: initialize IOAPIC and timers after ACPI

The proper order of initialization is: ACPI -> IOAPIC -> Timers.
The ACPI provided MADT table provides configuration for IOAPICs
and for example HPET.
Also, initialized IOAPIC is required by timers to setup their IRQs.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
arch,entry: cleanup _(from/to)_usermode macros

Add generic _from_usermode and _to_usermode macros that switch pagetable,
swap GS and switch stack.
Use these macros in syscall_(from/to)_usermode macros directly.
Rename enter_(from/to)_usermode macros to exc_(from/to)_usermode macros
to highlight the exception relation of theirs.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
arch,entry: add corrected irq_(from/to)_usermode macros

Previous use of old enter_(from/to)_usermode macros was incorrect as
it assumed the same offset of CS on the stack frame (but it was correct
for exceptions, where error_code+vector were also stored). Use cpu_irq_t
relative offset now.

Also, preserve USER_CS onto the switched stack when from usermode, to
let the to usermode macro properly detect the entry point.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
arch,entry: mask user flags in exception and irq handling

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
grub,unittest: enable timers for unit testing

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -109,9 +119,8 @@ ENTRY(handle_exception)
-
-    enter_to_usermode
+    exc_to_usermode
+    add $cpu_exc_ip, %_ASM_SP
```
I see what you're trying to do, but reading `cpu_exc_ip` feels odd.

Can you please add a comment that we just want to pop off `error_code` and `vector` from the stack (which is 8 bytes in any case, regardless of 32bit and 64bit, that's why I like it better, but anyway) and use the offset of `_ASM_IP` in `struct cpu_exc` to do this somehow arch-agnostic?

wipawel replied:
Fixed in updated patchset.

minipli-oss commented on include/arch/x86/processor.h:
```diff
@@ -245,18 +245,29 @@ typedef uint64_t x86_reg_t;
+
+    cpu_irq_t irq;
+} __packed;
 typedef struct cpu_exc cpu_exc_t;
```
This whole commit feels like a lot of code churn for no real gain :(

wipawel replied:
Commit and changes removed.

minipli-oss commented on common/setup.c:
```diff
@@ -241,11 +238,14 @@ void __noreturn __text_init kernel_start(uint32_t multiboot_magic, unsigned long
 
-    init_ioapic();
-
     init_pci();
```
Can you please explain (in the commit log) why that specific order is needed, e.g. why the IOAPIC needs to be initialized before the HPET or PIC? It feels counter-intuitive.

wipawel replied:
Done.

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -69,6 +69,25 @@
+        /* Preserve USER CS and a fake RIP on kernel stack */
+        /* This is expected by the irq_to_usermode         */
+        sub $cpu_irq_flags, %_ASM_SP
+        movw $__USER_CS, cpu_irq_cs(%_ASM_SP)
```
Ugh! Can't we simply mimic a regular exception behaviour and push a zero error code and the IRQ's vector number to have a common stack frame (mod in-kernel IRQs and exceptions that lack the RSP+SS part but that's not really needed).

wipawel replied:
Very good idea. Implemented in the new patchset.

minipli-oss reviewed (commented):
Can you please explain what's "completely broken"?

wipawel committed:
arch,x86: fix cs/ss padding for 32bit on stack irq/exc

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
entry: always push error_code and vector as machine words

By treating both error_code and vector as machines words (either
32bit or 64bit based on built architecture) we simplify
handling of the exceptions across architectures.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
arch,traps: cleanup IRQ vectors naming

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
asm-offsets: cleanup order of generated header

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
irq: make irq frame layout same as the exception one

On IRQ entry always push fake error_code value (always zero) and a
corresponding IRQ vector number on the stack to make its layout
match the exception stack frame layout. This simplifies the handling
of both exceptions and interrupts.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -32,70 +32,100 @@
+        /* This is expected by the irq_to_usermode to detect context          */
+        push $__USER_CS
+        sub $cpu_exc_cs, %_ASM_SP
+    1:
```
Why isn't `irq_from_usermode` the same as `exc_from_usermode`? It should be, no? The IRQ should already have switched to the kernel stack and pushed the exception frame there, so only the zero error code and vector number are needed to be able to make use of `exc_from_usermode` for the remainder. And that's what `asm_interrupt_handler_\sym` does so I'm puzzled why we need another macro that does something completely different -- and even has some hacks to fool the exit code.

wipawel replied:
The main reason is we do not use IST for anything but #DF. So for exceptions the stack stays the same for both user and kernel. But in this case it is fine, since exception for usermode is supposed to terminate the task. So for simplicity, we piggy back on the user stack, handle the `do_exception()` stuff and terminate it. And we do so, cause CPU put the exception frame on user stack already and we just want to use it.
For IRQs we do switch the stack for user, hence the different macro set. The CPU put the irq frame onto user stack, so we make it similar to the exception frame, switch to kernel (and use the hack to let the `irq_to_usermode()` detect the context in case of the user transition), handle the interrupt and switch back for the IRET.

wipawel replied:
Naaah, Mathias is right. The CPU uses TSS's RSP0 when IST is not specified. Just verified empirically. Going to make the changes much simpler now.

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -32,70 +32,100 @@
-#else
     push $\vec
-#endif
 
```
Yeah, this looks much cleaner, thanks! :)

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -32,70 +32,100 @@
-    MASK_USER_FLAGS
+    push $0 /* fake error_code */
+    push $\vec /* IRQ vector */
+    irq_from_usermode
```
Why not simply `exc_from_usermode`?

wipawel replied:
We do not use IST for irqs (yet). Please see the explanation above.

minipli-oss commented on include/arch/x86/processor.h:
```diff
@@ -201,15 +201,21 @@
+#elif defined(__i386__)
+typedef uint32_t x86_reg_t;
+#endif
+
```
I guess we cannot simplify it to something like this:
```c
typedef unsigned long x86_reg_t;
```
because we also support Windows as a build host which uses an LLP64 data model instead of the more common LP64?

wipawel replied:
Hmm... the reason for it was probably explicit highlighting on the size. I have no string opinion about it, so we can simplify it I guess.

minipli-oss commented on include/arch/x86/processor.h:
```diff
@@ -201,15 +201,21 @@
     struct __packed {
-        unsigned int E : 1, TLB : 2, index : 13, rsvd_sel : 16;
+        x86_reg_t E : 1, TLB : 2, index : 13;
     };
```
Uhm, I believe you need to specify the padding bits still which gets a little awkward when we want to support 32-/64-bit types. Otherwise it's uncertain in what order and where those bits will be allocated.

[gcc's documentation](https://gcc.gnu.org/onlinedocs/gcc/Structures-unions-enumerations-and-bit-fields-implementation.html) is a little scarce about it:

> * The order of allocation of bit-fields within a unit (C90 6.5.2.1, C99 and C11 6.7.2.1).
> 
> Determined by ABI.

I, myself, can't read anything out of this, beside, "if it breaks, you're free to keep all pieces :P".

Maybe we should make it portable and simply use masks to access the individual bits?

wipawel replied:
Yes, will fix it. But I really do not want to use the masks :-).

minipli-oss commented on include/arch/x86/processor.h:
```diff
@@ -201,15 +201,21 @@
     struct __packed {
-        unsigned int P : 1, W : 1, U : 1, R : 1, I : 1, rsvd_pfec : 27;
+        x86_reg_t P : 1, W : 1, U : 1, R : 1, I : 1;
     };
```
Same reasoning as for the bit-field above.

minipli-oss commented on include/arch/x86/processor.h:
```diff
@@ -239,23 +245,19 @@ extern void entry_SE(void);
-    uint32_t vector;
+    x86_reg_t vector;
+    x86_ex_error_code_t error_code;
 
```
Finally the correct order! :D

sktt commented on arch/x86/entry.S:
```diff
@@ -32,38 +32,46 @@
-        swapgs
+        _from_usermode switch_stack=0
     1:
 .endm
```
hm this reads a bit funny, because we're not necessarily coming from usermode. it's more like `maybe_exc_from_usermode`... Does it make sense to move the test/jz out from this?

wipawel replied:
Yes, I think we can wrap it up into a more explicit abstraction.

sktt commented on arch/x86/entry.S:
```diff
@@ -32,38 +32,46 @@
-
-.macro enter_to_usermode
-    _handle_usermode user_cr3
+.macro exc_to_usermode
+    testb $0x3, cpu_exc_cs(%_ASM_SP)
+    jz 1f /* skip if from kernel mode */
+        _to_usermode switch_stack=0
+    1:
```
Isn't this always executed in kernel mode? Again does it make sense to hoist the conditional?

edit: yea i see, it's whether the interrupt triggered in kernel mode.. Maybe `exc_cond_{to,from}_usermode`

wipawel replied:
Yes. We have to be sure whether we did the transition earlier or not. Will apply some abstraction to it.

sktt commented on arch/x86/entry.S:
```diff
@@ -86,7 +95,7 @@ END_FUNC(entry_\sym)
 .macro interrupt_handler sym func
 ENTRY(asm_interrupt_handler_\sym)
-    enter_from_usermode
+    _from_usermode
```
Is this always triggered in user mode? What happens if interrupted in kernel mode?

wipawel replied:
It can be triggered in kernel mode too. The code should use irq_from_usermode here.

sktt reviewed (commented):
I am not 100% confident about my comments here so please let me know if my comments are valid or not haha 

the main concern thing is that I think that semantically `exc_{to,from}_usermode` should only be used when actually going to/from usermode.

sktt commented on grub/grub-test.cfg:
```diff
@@ -10,6 +10,6 @@ terminal_input --append serial
 
 menuentry "kernel64" {
-   multiboot2 /boot/kernel64.bin.xz poweroff=1 com1=0x3f8,115200,8,n,1      integer=42   boolean=1   string=foo badstring=toolong booleantwo tests=unit_tests
+   multiboot2 /boot/kernel64.bin.xz poweroff=1 com1=0x3f8,115200,8,n,1 pit=1 hpet=1 apic_timer=1     integer=42   boolean=1   string=foo badstring=toolong booleantwo tests=unit_tests
```
unittest expects these

wipawel commented:
> Can you please explain what's "completely broken"?

Turns out, my understanding of interrupt context switching in long mode was "completely broken". Thanks!
